resources/channel.sub.js

```diff
@@ -0,0 +1,1003 @@
+                        typeof self[item.value.type] === "function") {
+                        result = new self[item.value.type](item.value.message);
+                    } else {
+                        result = new Error(item.value.message);
+                    }
+                    result.name = item.value.name;
+                    result.lineNumber = item.value.lineNumber;
+                    result.columnNumber = item.value.columnNumber;
+                    result.fileName = item.value.fileName;
+                    result.stack = item.value.stack;
+                    break;
+                case "array":
+                    result = [];
+                    newTarget = {type: "array", value: result};
+                    for (let child of value) {
+                        stack.push({item: child, target: newTarget});
+                    }
+                    break;
+                case "set":
+                    result = new Set();
+                    newTarget = {type: "set", value: result};
+                    for (let child of value) {
+                        stack.push({item: child, target: newTarget});
+                    }
+                    break;
+                case "object":
+                    result = {};
+                    newTarget = {type: "object", value: result};
+                    for (let [targetKey, child] of Object.entries(value)) {
+                        stack.push({item: child, target: newTarget, targetKey});
+                    }
+                    break;
+                case "map":
+                    result = new Map();
+                    newTarget = {type: "map", value: result};
+                    for (let [key, child] of value) {
+                        stack.push({item: key, target: newTarget, targetKey: "key"});
+                        stack.push({item: child, target: newTarget, targetKey: "value"});
+                    }
+                    break;
+                default:
+                    throw new TypeError(`Can't deserialize object of type ${type}`);
+                }
+                if (objectId !== undefined) {
+                    objectMap.set(objectId, result);
+                }
+            }
+
+            if (target === null) {
+                if (deserialized !== null) {
+                    throw new Error("Tried to create multiple output values");
+                }
+                deserialized = result;
+            } else {
+                switch(target.type) {
+                case "array":
+                    target.value.push(result);
+                    break;
+                case "set":
+                    target.value.add(result);
+                    break;
+                case "object":
+                    target.value[targetKey] = result;
+                    break;
+                case "map":
+                    if (targetKey === "key") {
+                        target.key = result;
```

Comment:
Stuffing a value on `target` here makes me wonder. Like with serialization, it depends on the order that the queue is managed. We might just as well use a local variable, making it more clear that we only store one key at a time, and that things would blow up the order were different.

Reply:
Oh yeah, you're right!